### PR TITLE
Prevent automatic footer refreshes from drawing the footer on top of stuff that isn't ReaderUI

### LINKED ISSUE #6616: Refreshing time shows up minibar
Kindle PW2, 2020.08.1-44

Refreshing time shows up minibar on pages where it shouldn't be shown (History, TOC, popup footnotes)

### PR DESCRIPTION
**Pull request by NiLuJe**
Issue exposed by #6540, as long as `auto_refresh_time` is enabled.

Fix #6616

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/6648)

### CODE REVIEW

**Comment by NiLuJe** [comment]
I may have missed a few things that could potentially need a `covers_footer` flag, but I can't think of any others right now ;).

**Comment by NiLuJe** [comment]
I'm wary of switching to a purely heuristics-based approach, because the few ideas proposed have potential issues:

* I'm not sure if we can assume the stack size to be constant, so, checking that basically *only* ReaderUI is on the stack may not work all that well...
* Doing a hit check based on region like @poire-z proposed may not always work either, because regions are often unreliable or hard to compute.

**Review by Frenzie — APPROVED** [review]
Seems okay, but it feels a bit overly complex/possibly the wrong place?

**Comment by NiLuJe** [comment]
I could technically move most of it to a specialized `UIManager:widgetRepaint` variant, but it's be a pretty specific variant, since the `covers_footer` stuff pretty much limits it to, well, the footer ;).

File: frontend/apps/reader/modules/readerfooter.lua [hunk 1]
```diff
@@ -2005,6 +2029,13 @@ function ReaderFooter:onSuspend()
     end
 end
 
+-- We want to be able to disable auto_refresh_time when displaying *some* non-fullscreen widgets on top of ReaderUI,
+-- otherwise it'll happily keep on ticking and drawing on top of stuff it ought not to... (#6616)
+--- @note: If the widget actually makes it to UIManager's stack (i.e., it's passed to UIManager:show()),
+-- it's generally simpler to set covers_footer when initializing the Widget object.
+ReaderFooter.onDisableFooterAutoRefresh = ReaderFooter.onSuspend
+ReaderFooter.onRestoreFooterAutoRefresh = ReaderFooter.onResume
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerfooter.lua:2037]: Looks like you're no more using these ?

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerfooter.lua:2037]: Currently, no, but they may come in handy if we need to do this for subwidgets that never make it to the stack (i.e., never passed to `UIManager:show`).

Which, granted, will potentially never happen ^^.

I guess I could leave it, but commented out, then?

Comment (poire-z) [reply frontend/apps/reader/modules/readerfooter.lua:2037]: Yes, I'd rather see them commented out, so I won't wonder why/how these are used if I ever look at that footer again :)

**Comment by NiLuJe** [comment]
> I could technically move most of it to a specialized `UIManager:widgetRepaint` variant, but it'd be a pretty specific variant, since the `covers_footer` stuff pretty much limits it to, well, the footer ;).

Actually, while I can do that, that still generates a bogus setDirty call :/.

```diff
diff --git a/frontend/apps/reader/modules/readerfooter.lua b/frontend/apps/reader/modules/readerfooter.lua
index 2107a39d..4954189e 100644
--- a/frontend/apps/reader/modules/readerfooter.lua
+++ b/frontend/apps/reader/modules/readerfooter.lua
@@ -19,7 +19,6 @@ local UIManager = require("ui/uimanager")
 local VerticalGroup = require("ui/widget/verticalgroup")
 local VerticalSpan = require("ui/widget/verticalspan")
 local WidgetContainer = require("ui/widget/container/widgetcontainer")
-local logger = require("logger")
 local util = require("util")
 local T = require("ffi/util").template
 local _ = require("gettext")
@@ -562,30 +561,7 @@ end
 function ReaderFooter:setupAutoRefreshTime()
     if not self.autoRefreshTime then
         self.autoRefreshTime = function()
-            -- Don't refresh if there's another widget than ReaderUI flagged as covers_fullscreen being shown,
-            -- or if a non-fullscreen widget is flagged as covers_footer.
-            -- This isn't particularly pretty, but, oh, well (#6616).
-            local skip_refresh = false
-            -- c.f., UIManager:_repaint
-            for i = #UIManager._window_stack, 1, -1 do
-                local widget = UIManager._window_stack[i].widget
-                if widget.covers_fullscreen then
-                    if widget.name and widget.name == "ReaderUI" then  -- luacheck: ignore
-                        -- NOP
-                    else
-                        skip_refresh = true
-                        logger.dbg("Skipping ReaderFooter:autoRefreshTime refresh, because something covers ReaderUI")
-                        break
-                    end
-                elseif widget.covers_footer then
-                        skip_refresh = true
-                        logger.dbg("Skipping ReaderFooter:autoRefreshTime refresh, because something covers ReaderFooter")
-                        break
-                end
-            end
-            if not skip_refresh then
-                self:onUpdateFooter(true)
-            end
+            self:onUpdateFooter(true)
             UIManager:scheduleIn(61 - tonumber(os.date("%S")), self.autoRefreshTime)
         end
     end
@@ -1822,8 +1798,7 @@ function ReaderFooter:_updateFooterText(force_repaint, force_recompute)
             UIManager:setDirty(self.view.footer, function()
                 return "ui", refresh_dim
             end)
-            -- c.f., ReaderView:paintTo()
-            UIManager:widgetRepaint(self.view.footer, 0, 0)
+            UIManager:repaintReaderFooter(self.view.footer)
         else
             UIManager:setDirty(self.view.dialog, function()
                 return "ui", refresh_dim
diff --git a/frontend/ui/uimanager.lua b/frontend/ui/uimanager.lua
index 77e6a961..ba678e40 100644
--- a/frontend/ui/uimanager.lua
+++ b/frontend/ui/uimanager.lua
@@ -1141,6 +1141,40 @@ function UIManager:widgetRepaint(widget, x, y)
     widget:paintTo(Screen.bb, x, y)
 end
 
+-- Now, this one is basically widgetRepaint, but tailored for a very,
+-- very specific use-cases related to ReaderFooter:setupAutoRefreshTime
+-- (c.f., #6648)
+function UIManager:repaintReaderFooter(widget)
+    if not widget then return end
+
+    -- Don't repaint if there's another widget than ReaderUI flagged as covers_fullscreen being shown,
+    -- or if a non-fullscreen widget is flagged as covers_footer.
+    -- This isn't particularly pretty, but, oh, well (#6616).
+    local skip_repaint = false
+    -- c.f., UIManager:_repaint
+    for i = #self._window_stack, 1, -1 do
+        local widget = self._window_stack[i].widget
+        if widget.covers_fullscreen then
+            if widget.name and widget.name == "ReaderUI" then  -- luacheck: ignore
+                -- NOP
+            else
+                skip_repaint = true
+                logger.dbg("Skipping ReaderFooter repaint, because something covers ReaderUI")
+                break
+            end
+        elseif widget.covers_footer then
+            skip_repaint = true
+            logger.dbg("Skipping ReaderFooter repaint, because something covers ReaderFooter")
+            break
+        end
+    end
+    if skip_repaint then return end
+
+    logger.dbg("Explicit ReaderFooter repaint:", widget.name or widget.id or tostring(widget))
+    -- c.f., ReaderView:paintTo()
+    widget:paintTo(Screen.bb, 0, 0)
+end
+
 function UIManager:setInputTimeout(timeout)
     self.INPUT_TIMEOUT = timeout or 200*1000
 end
```

**Comment by NiLuJe** [comment]
> > I could technically move most of it to a specialized `UIManager:widgetRepaint` variant, but it'd be a pretty specific variant, since the `covers_footer` stuff pretty much limits it to, well, the footer ;).
> 
> Actually, while I can do that, that still generates a bogus setDirty call :/.

Okay, that was trivial to fix ;).

**Review by poire-z — APPROVED** [review]

**Comment by gerhaher** [comment]
Problem still happens with Dictionary (Large window).

**Comment by NiLuJe** [comment]
Yeah, I had an idea on how to deal with that last night, will check later ;).

**Pull request merged**